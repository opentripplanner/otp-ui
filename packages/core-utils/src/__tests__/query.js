import clone from "lodash/cloneDeep";

import {
  restoreDateNowBehavior,
  setDefaultTestTime
} from "../../../../test-utils/time";

import {
  getDefaultQuery,
  getRoutingParams,
  isNotDefaultQuery,
  parseLocationString,
  planParamsToQuery
} from "../query";

const config = require("./__mocks__/config.json");

describe("query", () => {
  afterEach(restoreDateNowBehavior);

  describe("getDefaultQuery", () => {
    it("should return default query", () => {
      setDefaultTestTime();
      expect(getDefaultQuery()).toMatchSnapshot();
    });

    it("should allow config overrides for default query", () => {
      setDefaultTestTime();
      const configWithQueryParamOverrides = {
        defaultQueryParams: {
          maxWalkDistance: 3219 // 2 miles
        },
        routingTypes: [{ key: "ITINERARY", text: "Exact time" }]
      };
      expect(getDefaultQuery(configWithQueryParamOverrides)).toMatchSnapshot();
    });
  });

  describe("getRoutingParams", () => {
    // the config isn't actually used except for some wheelchair check, so it is
    // ok to send an empty object for most tests
    const fakeConfig = {};
    const makeBaseTestQuery = () => ({
      ...getDefaultQuery(),
      date: "2020-08-24",
      from: {
        lat: 36.95471026341096,
        lon: -122.0248185852425,
        name: "Steamer Lane, Santa Cruz, CA, 95060, USA"
      },
      time: "21:53",
      to: {
        lat: 36.961843106786766,
        lon: -122.02402657342725,
        name: "Municipal Wharf St, Santa Cruz, CA, 95060, USA"
      }
    });
    it("should create routing params for transit query", () => {
      expect(
        getRoutingParams(fakeConfig, makeBaseTestQuery())
      ).toMatchSnapshot();
    });

    it("should create routing params for car rental query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          companies: "FAKECOMPANY",
          mode: "CAR_RENT,BUS"
        })
      ).toMatchSnapshot();
    });

    it("should create routing params for personal micromobility query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          companies: "FAKECOMPANY",
          mode: "MICROMOBILITY,BUS",
          watts: 250
        })
      ).toMatchSnapshot();
    });

    it("should create routing params for intermediate places query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          intermediatePlaces: [
            {
              lat: 36.95145225211049,
              lon: -122.02672866668283,
              name: "Lighthouse Point Park, Santa Cruz, CA, USA"
            },
            {
              lat: 36.95028619314903,
              lon: -122.05711810284686,
              name:
                "Natural Bridges State Beach, 2531 W Cliff Dr, Santa Cruz, CA, 95060, USA"
            }
          ]
        })
      ).toMatchSnapshot();
    });

    it("should create routing params for standard flex query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode: "WALK,TRANSIT,FLEX"
        })
      ).toMatchSnapshot();
    });

    it("should create repaired routing params for broken flex query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode: "WALK,FLEX_DIRECT,TRANSIT,FLEX"
        })
      ).toMatchSnapshot();
    });

    it("should create routing params for flex-only query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode: "FLEX"
        })
      ).toMatchSnapshot();
    });

    it("should create routing params for standard flex query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode: "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT"
        })
      ).toMatchSnapshot();
    });

    it("should create repaired routing params for very broken flex query", () => {
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode: "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT,FLEX,WALK,BIKE"
        })
      ).toMatchSnapshot();
      expect(
        getRoutingParams(fakeConfig, {
          ...makeBaseTestQuery(),
          mode:
            "FLEX_EGRESS,FLEX,FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT,FLEX,WALK,BIKE,WALK,TRANSIT,BUS"
        })
      ).toMatchSnapshot();
    });
  });

  describe("isNotDefaultQuery", () => {
    it("should return false for default query", () => {
      setDefaultTestTime();
      expect(isNotDefaultQuery(getDefaultQuery(config), config)).toBe(false);
    });

    it("should return false for default query with config overrides", () => {
      // Clone default config and add new default maxWalkDistance
      const configWithQueryParamOverrides = clone(config);
      configWithQueryParamOverrides.defaultQueryParams = {
        maxWalkDistance: 3219
      };
      setDefaultTestTime();
      expect(
        isNotDefaultQuery(
          getDefaultQuery(configWithQueryParamOverrides),
          configWithQueryParamOverrides
        )
      ).toBe(false);
    });

    it("should return true for query with modified maxWalkDistance", () => {
      setDefaultTestTime();
      const query = getDefaultQuery(config);
      // Double the max walk distance
      query.maxWalkDistance *= 2;
      expect(isNotDefaultQuery(query, config)).toBe(true);
    });

    it("should return true for query with non-default mode", () => {
      setDefaultTestTime();
      const query = getDefaultQuery(config);
      // Set the mode to walk
      query.mode = "WALK";
      expect(isNotDefaultQuery(query, config)).toBe(true);
    });
  });

  describe("parseLocationString", () => {
    it("should return null for null input", () => {
      expect(parseLocationString(null)).toMatchSnapshot();
    });

    it("should return location for valid input", () => {
      expect(
        parseLocationString("123 Main St::33.983929829,-87.3892387982")
      ).toMatchSnapshot();
    });

    it("should return location with coordinates as name for coordinates-only input", () => {
      expect(
        parseLocationString("33.983929829,-87.3892387982")
      ).toMatchSnapshot();
    });
  });

  describe("planParamsToQuery", () => {
    const makeBaseTestQuery = () => ({
      arriveBy: "false",
      bannedRoutes: "897ABC",
      companies: "",
      date: "2019-10-31",
      fromPlace:
        "Guide Dogs for the Blind, Portland, OR, USA::45.517373,-122.675601",
      ignoreRealtimeUpdates: "true",
      maxWalkDistance: "1207",
      mode: "BUS,TRAM,RAIL,GONDOLA,WALK",
      optimize: "QUICK",
      showIntermediateStops: "true",
      time: "17:45",
      toPlace: "Weather Machine, Portland, OR, USA::45.519015,-122.679321",
      ui_activeItinerary: "0",
      ui_activeSearch: "irc7h4rb8",
      walkSpeed: "1.34"
    });

    it("should parse a depart at query", () => {
      expect(planParamsToQuery(makeBaseTestQuery())).toMatchSnapshot();
    });

    it("should parse a query with intermediate places", () => {
      expect(
        planParamsToQuery({
          ...makeBaseTestQuery(),
          intermediatePlaces: [
            "Mill Ends Park, Downtown - Portland::45.516202,-122.673247",
            "International Rose Test Garden, Portland::45.519094,-122.705692"
          ]
        })
      ).toMatchSnapshot();
    });
  });
});
