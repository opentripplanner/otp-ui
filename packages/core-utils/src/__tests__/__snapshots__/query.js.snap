// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query getDefaultQuery should allow config overrides for default query 1`] = `
Object {
  "bikeSpeed": 3.58,
  "date": "2019-08-04",
  "departArrive": "NOW",
  "endTime": "09:00",
  "from": null,
  "ignoreRealtimeUpdates": false,
  "intermediatePlaces": Array [],
  "maxBikeDistance": 4828,
  "maxBikeTime": 20,
  "maxEScooterDistance": 4828,
  "maxWalkDistance": 3219,
  "maxWalkTime": 15,
  "mode": "WALK,TRANSIT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "optimizeBike": "SAFE",
  "otherThanPreferredRoutesPenalty": 900,
  "routingType": "ITINERARY",
  "showIntermediateStops": true,
  "startTime": "07:00",
  "time": "19:34",
  "to": null,
  "walkSpeed": 1.34,
  "watts": 250,
  "wheelchair": false,
}
`;

exports[`query getDefaultQuery should return default query 1`] = `
Object {
  "bikeSpeed": 3.58,
  "date": "2019-08-04",
  "departArrive": "NOW",
  "endTime": "09:00",
  "from": null,
  "ignoreRealtimeUpdates": false,
  "intermediatePlaces": Array [],
  "maxBikeDistance": 4828,
  "maxBikeTime": 20,
  "maxEScooterDistance": 4828,
  "maxWalkDistance": 1207,
  "maxWalkTime": 15,
  "mode": "WALK,TRANSIT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "optimizeBike": "SAFE",
  "otherThanPreferredRoutesPenalty": 900,
  "routingType": "ITINERARY",
  "showIntermediateStops": true,
  "startTime": "07:00",
  "time": "19:34",
  "to": null,
  "walkSpeed": 1.34,
  "watts": 250,
  "wheelchair": false,
}
`;

exports[`query getRoutingParams should create repaired routing params for broken flex query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "maxWalkDistance": 1207,
  "mode": "WALK,FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT,TRANSIT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query getRoutingParams should create repaired routing params for very broken flex query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "mode": "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT,WALK,BIKE",
  "numItineraries": 3,
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query getRoutingParams should create repaired routing params for very broken flex query 2`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "maxWalkDistance": 1207,
  "mode": "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT,WALK,BIKE,WALK,TRANSIT,BUS",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query getRoutingParams should create routing params for car rental query 1`] = `
Object {
  "arriveBy": false,
  "companies": "FAKECOMPANY",
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "minTransitDistance": "50%",
  "mode": "CAR_RENT,BUS,WALK",
  "numItineraries": 3,
  "onlyTransitTrips": true,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "searchTimeout": 10000,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
}
`;

exports[`query getRoutingParams should create routing params for flex-only query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "mode": "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT",
  "numItineraries": 3,
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
}
`;

exports[`query getRoutingParams should create routing params for intermediate places query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "intermediatePlaces": Array [
    "Lighthouse Point Park, Santa Cruz, CA, USA::36.95145225211049,-122.02672866668283",
    "Natural Bridges State Beach, 2531 W Cliff Dr, Santa Cruz, CA, 95060, USA::36.95028619314903,-122.05711810284686",
  ],
  "maxWalkDistance": 1207,
  "mode": "WALK,TRANSIT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query getRoutingParams should create routing params for personal micromobility query 1`] = `
Object {
  "arriveBy": false,
  "companies": "FAKECOMPANY",
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "maxEScooterDistance": 4828,
  "maxWalkDistance": 4828,
  "maximumMicromobilitySpeed": 5,
  "mode": "MICROMOBILITY,BUS",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "watts": 250,
  "weight": 102.5,
}
`;

exports[`query getRoutingParams should create routing params for standard flex query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "maxWalkDistance": 1207,
  "mode": "WALK,TRANSIT,FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query getRoutingParams should create routing params for standard flex query 2`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "mode": "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT",
  "numItineraries": 3,
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
}
`;

exports[`query getRoutingParams should create routing params for transit query 1`] = `
Object {
  "arriveBy": false,
  "date": "2020-08-24",
  "fromPlace": "Steamer Lane, Santa Cruz, CA, 95060, USA::36.95471026341096,-122.0248185852425",
  "ignoreRealtimeUpdates": false,
  "maxWalkDistance": 1207,
  "mode": "WALK,TRANSIT",
  "numItineraries": 3,
  "optimize": "QUICK",
  "otherThanPreferredRoutesPenalty": 900,
  "showIntermediateStops": true,
  "time": "21:53",
  "toPlace": "Municipal Wharf St, Santa Cruz, CA, 95060, USA::36.961843106786766,-122.02402657342725",
  "walkSpeed": 1.34,
}
`;

exports[`query parseLocationString should return location for valid input 1`] = `
Object {
  "lat": 33.983929829,
  "lon": -87.3892387982,
  "name": "123 Main St",
}
`;

exports[`query parseLocationString should return location with coordinates as name for coordinates-only input 1`] = `
Object {
  "lat": 33.983929829,
  "lon": -87.3892387982,
  "name": "33.98393, -87.38924",
}
`;

exports[`query parseLocationString should return null for null input 1`] = `null`;

exports[`query planParamsToQuery should parse a depart at query 1`] = `
Object {
  "bannedRoutes": "897ABC",
  "companies": "",
  "date": "2019-10-31",
  "departArrive": "DEPART",
  "from": Object {
    "lat": 45.517373,
    "lon": -122.675601,
    "name": "Guide Dogs for the Blind, Portland, OR, USA",
  },
  "ignoreRealtimeUpdates": "true",
  "maxWalkDistance": 1207,
  "mode": "BUS,TRAM,RAIL,GONDOLA,WALK",
  "optimize": "QUICK",
  "showIntermediateStops": "true",
  "time": "17:45",
  "to": Object {
    "lat": 45.519015,
    "lon": -122.679321,
    "name": "Weather Machine, Portland, OR, USA",
  },
  "ui_activeItinerary": 0,
  "ui_activeSearch": "irc7h4rb8",
  "walkSpeed": 1.34,
}
`;

exports[`query planParamsToQuery should parse a query with intermediate places 1`] = `
Object {
  "bannedRoutes": "897ABC",
  "companies": "",
  "date": "2019-10-31",
  "departArrive": "DEPART",
  "from": Object {
    "lat": 45.517373,
    "lon": -122.675601,
    "name": "Guide Dogs for the Blind, Portland, OR, USA",
  },
  "ignoreRealtimeUpdates": "true",
  "intermediatePlaces": Array [
    Object {
      "lat": 45.516202,
      "lon": -122.673247,
      "name": "Mill Ends Park, Downtown - Portland",
    },
    Object {
      "lat": 45.519094,
      "lon": -122.705692,
      "name": "International Rose Test Garden, Portland",
    },
  ],
  "maxWalkDistance": 1207,
  "mode": "BUS,TRAM,RAIL,GONDOLA,WALK",
  "optimize": "QUICK",
  "showIntermediateStops": "true",
  "time": "17:45",
  "to": Object {
    "lat": 45.519015,
    "lon": -122.679321,
    "name": "Weather Machine, Portland, OR, USA",
  },
  "ui_activeItinerary": 0,
  "ui_activeSearch": "irc7h4rb8",
  "walkSpeed": 1.34,
}
`;
